### PR TITLE
CI: add pre-commit.ci to apply clang-format and clang-tidy fixes

### LINKED ISSUE #3915: clang-format and clang-tidy is not employed
### Describe the Code Quality Issue

clang-format can format c++ source files; clang-tidy is used to perform static analysis, helping improve code quality.
Adding CI checks utilizing the existed `.clang-tidy` and `.clang-format` in the repo on the new-coming PRs would be better.
Some third-party add-ons like https://pre-commit.ci/ is also helpful.

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Identify the specific code file or section with the code quality issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

### PR DESCRIPTION
**Pull request by caic99**
### Linked Issue
Fix #3915

### What's changed?

This PR add the CI test static checks using `clang-format` and `clang-tidy`. It uses pre-commit.ci to orchestrate the checks and submit patches. On new commits pushing, it will try to fix the code smells in changed files.

For example, a developer modified the code without caring indentations in [c17732b](https://github.com/deepmodeling/abacus-develop/pull/4225/commits/c17732b5690e2c7d3a71a032226019ffb2aed0e4). The commit triggered workflow runs, and pre-commit.ci submited a patch for it in [039d5a0](https://github.com/deepmodeling/abacus-develop/pull/4225/commits/039d5a096cd6259d15a870e7c7bab11cfc633589).

This PR also updates the config files used by those tools. Please submit an issue if the suggested changes by those rules does not fit the circumstance well.

### CODE REVIEW

File: source/driver.h [hunk 1]
```diff
@@ -19,12 +19,12 @@ class Driver
      * This function read the parameter in "INPUT", "STRU" etc,
      * and split the MPI world into different groups.
      */
-    void reading();
+    static void reading();
```
Comment (dyzheng) [inline source/driver.h:22]: Why do you add static for this function?

Comment (caic99) [reply source/driver.h:22]: This is a check in clang-tidy. Please see https://clang.llvm.org/extra/clang-tidy/checks/readability/convert-member-functions-to-static.html

**Review by dyzheng — APPROVED** [review]
LGTM, but I want to know how to make the decision to change `clang-format` in the future?

**Comment by caic99** [comment]
> LGTM, but I want to know how to make the decision to change `clang-format` in the future?

@dyzheng As I've mentioned, one can change the check rules if current rules mismatch their requirements. It is possible to polish the rules iteratively.

**Comment by mohanchen** [comment]
Great job! I need to check every rule in the clang-format and clang-tidy before merging this PR.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+fail_fast: false
+repos:
+  - repo: https://github.com/pocc/pre-commit-hooks
+    rev: v1.3.5
+    hooks:
+      - id: clang-format
+        args: [-i]
+      - id: clang-tidy
+        args: [-p=build, --fix-errors]
```
Comment (caic99) [inline .pre-commit-config.yaml]: Disable clang-tidy checks before @mohanchen has the rules reviewed.
```suggestion
      # - id: clang-tidy
      #   args: [-p=build, --fix-errors]
```

**Review by mohanchen — APPROVED** [review]

**Pull request merged**